File: pydrpc/drpc_client.py
import socket
import time

DEFAULT_TIMEOUT = 0.5


class drpc_arg_wrapper:
    args = None
    size: int = 0

    def __init__(self, args):
        self.args = args
        self.size = len(bytes(self.args))

    def __bytes__(self) -> bytes:
        return self.size.to_bytes(8, "big") + bytes(self.args)

    def __repr__(self) -> str:
        return f'{bytes(self.args)}\n{self.size}'
    pass


class drpc_msg:
    target: str = ""
    req: drpc_arg_wrapper = None
    rep: drpc_arg_wrapper = None
    pass


class drpc_host:
    port = 0
    hostname = ""
    pass


class drpc_client:
    def __init__(self, timeout_val=DEFAULT_TIMEOUT):
        self.timeout_val = timeout_val
        pass

    def Call(self, dh: drpc_host, m: drpc_msg) -> int:
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
            try:
                sock.connect((dh.hostname, dh.port))
            except:
                return -1
            # send RPC
            # target function
            target_str_len = len(m.target)
            sock.sendall(target_str_len.to_bytes(8, "little"))
            sock.sendall(m.target.encode("ascii"))
            # request args
            sock.sendall(m.req.size.to_bytes(8, "little"))
            sock.sendall(bytes(m.req.args))
            # reply
            sock.sendall(m.rep.size.to_bytes(8, "little"))
            sock.sendall(bytes(m.rep.args))
            # checksum
            # todo
            
            deadline = time.time() + self.timeout_val
            sock.settimeout(deadline - time.time())
            # receive RPC reply
            # reply
            try:
                b_size = sock.recv(8, socket.MSG_WAITALL)
                m.rep.size = int.from_bytes(b_size, "little")
                m.rep.args.serialize(sock.recv(m.rep.size, socket.MSG_WAITALL))
            except:
                pass

        return 0

File: src/drpc_client.cpp
#include "drpc.h"
#include <string>
#include <netdb.h> // gethostbyname(), struct hostent
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <cerrno>
#include <string.h>

drpc_client::drpc_client() : timeout_val(DEFAULT_TIMEOUT) {}

drpc_client::drpc_client(const int timeout) : timeout_val(timeout) {}

drpc_client::drpc_client(const int timeout, bool block) : timeout_val(timeout), blocking(block) {}

int drpc_client::Call(drpc_host &srv, std::string funct, rpc_arg_wrapper *args, rpc_arg_wrapper *err)
{
    drpc_msg m{funct, args, err};
    int status = do_rpc(srv, m);
    if (status != 0)
    {
        perror("unable to complete RPC");
        return -1;
    }
    return 0;
}

int drpc_client::do_rpc(drpc_host &srv, drpc_msg &m)
{
    // connect to drpc server port
    struct sockaddr_in addr;
    addr.sin_family = AF_INET;
    struct hostent *host = gethostbyname(srv.hostname);
    if (host == nullptr)
    {
        fprintf(stderr, "%s: unknown host\n", srv.hostname);
        return 1;
    }
    memcpy(&(addr.sin_addr), host->h_addr, host->h_length);
    addr.sin_port = htons(srv.port);
    int sockfd = socket(AF_INET, SOCK_STREAM, 0);
    if (sockfd < 0)
    {
        perror("Failed to create socket.");
        return 1;
    }

    struct timeval tv = {0, timeout_val * 1000}; // sleep for TIMEOUT ms

    if (!blocking)
    {
        if (setsockopt(sockfd, SOL_SOCKET, SO_RCVTIMEO, (char *)&tv, sizeof(tv)) < 0)
        {
            perror("Failed to set socket receive timeout.");
            return 1;
        }
    }

    if (connect(sockfd, (sockaddr *)&addr, sizeof(addr)) == -1)
    {
        perror("Error connecting stream socket");
        return 1;
    }

    // send RPC
    // target function
    secure_send(sockfd, (void*)m.target.c_str(), m.target.size());
    // request args
    secure_send(sockfd, m.req->args, m.req->len);
    // reply
    secure_send(sockfd, m.rep->args, m.rep->len);

    // receive RPC reply
    // reply
    void** data_ptr = &m.rep->args;
    secure_recv(sockfd, data_ptr);

    close(sockfd);

    return 0;
}